pyface/color.py

```diff
@@ -0,0 +1,260 @@
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" Color classes and corresponding trait types for Pyface.
+
+The base Color class holds red, green, blue and alpha channel values as
+a tuple of normalized values from 0.0 to 1.0.  Various property traits
+pull out the individual channel values and supply values for the HSV
+and HSL colour spaces (with and without alpha).
+
+The ``from_toolkit`` and ``to_toolkit`` methods allow conversion to and
+from native toolkit color objects.
+"""
+
+import colorsys
+
+from traits.api import (
+    HasStrictTraits, Property, Range, Tuple, cached_property
+)
+
+
+def channels_to_ints(channels, maximum=255):
+    """ Convert an iterable of floating point channel values to integers.
+
+    Values are rounded to the nearest integer, rather than truncated.
+
+    Parameters
+    ----------
+    channels : iterable of float
+        An iterable of channel values, each value between 0.0 and 1.0,
+        inclusive.
+    maximum : int
+        The maximum value of the integer range.  Common values are 15,
+        65535 or 255, which is the default.
+
+    Returns
+    -------
+    values : tuple of int
+        A tuple of values as integers between 0 and max, inclusive.
+    """
+    return tuple(int(round(channel * maximum)) for channel in channels)
+
+
+def ints_to_channels(values, maximum=255):
+    """ Convert an iterable of integers to floating point channel values.
+
+    Parameters
+    ----------
+    values : tuple of int
+        An iterable of values as integers between 0 and max, inclusive.
+    maximum : int
+        The maximum value of the integer range.  Common values are 15,
+        65535 or 255, which is the default.
+
+    Returns
+    -------
+    channels : iterable of float
+        A tuple of channel values, each value between 0.0 and 1.0,
+        inclusive.
+    """
+    return tuple(value / maximum for value in values)
+
+
+#: A trait holding a single channel value.
+Channel = Range(0.0, 1.0, value=1.0, channel=True)
+
+#: A trait holding three channel values.
+ChannelTuple = Tuple(Channel, Channel, Channel)
+
+#: A trait holding four channel values.
+AlphaChannelTuple = Tuple(Channel, Channel, Channel, Channel)
+
+
+class Color(HasStrictTraits):
+    """ A mutable specification of a color with alpha.
+
+    This is a class designed to be used by user interface elements which
+    need to color some or all of the interface element.  Each color has a
+    number of different representations as channel tuples, each channel
+    holding a value between 0.0 and 1.0, inclusive.  The standard red,
+    green, blue and alpha channels are also provided as convenience
+    properties.
+
+    Methods are provided to convert to and from toolkit-specific color
+    objects.
+
+    Colors implement equality testing, but are not hashable as they are
+    mutable, and so are not suitable for use as dictionary keys.  If you
```

Comment:
The objects that this is intended to replace (`QColor` and `wx.Colour` values of the `Color` trait) are mutable.

If we were to make this immutable, the design would be a lot different (probably a `namedtuple` rather than a `HasTraits`)

Reply:
I guess it depends on how this object is to be used. If we only ever interface it with the GUI toolkit using `from_toolkit` and `to_toolkit`, then we might not need to mutate the traits on this object at all.

Immutables are much easier to work with (and are much easier to maintain), especially in asynchronous context (GUI is such context). It is probably also easier to start with an immutable, and then turn it into a mutable later. And if later we run into a situation where this object really needs to be mutated, we have a reason to do it. On the other hand, it would be harder to make a mutable an immutable later.

I will leave that up to you whether you want to consider turning this an immutable.

Reply:
No, I expect that the to/from toolkit functions will be used at the point of need internally by toolkit-specific code and not exposed to typical users.  One goal of this exercise is that eventually we should never have someone have to use a `QColour` or `wx.Colour` unless they are writing toolkit-specific code.